kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/KeyContextAllocator.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.inband;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import javax.security.auth.DestroyFailedException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.filter.encryption.EncryptionException;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * The idea is the KeyContextAllocator becomes responsible allocating a DEK to
+ * a client that wants to encrypt N records (creating a new DEK if required). The CompletionStage
```

Comment:
That service exists, it's the `KeyManager`. But the implementation of `count the usages of the dek and thus switch DEK's as required` is a little challenging in the face of async code. Hoping that someone will point out a better way to implement this. Reference-counting feels too heavy.

Also makes me think that we need to separate something because the DEK creation and rotation work would likely be common to all KeyManagers. It shouldn't be a detail particular to the InBandKeyManager